site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+hostname. What semantics should apply when a `GRPCRoute` and an `HTTPRoute` both apply to
+a request?
+
+We propose that the GRPCRoute's matchers be translated into the equivalent
+HTTPRoute matchers and that the normal HTTPRoute precedence rules apply.
+This translation is mostly straightforward, but one significant deviation between
+GRPCRoute and HTTPRoute poses difficulties: the gRPC method matcher.
+
+gRPC methods are encoded in the URI of an HTTP/2 message in the format
+`/SERVICE/METHOD`. The matcher allows the user to supply these independently, so
+that the user may leave either blank, to be interpreted as an unconditional
+match. Further, this field may either be an exact match or a regex match.
+
+To translate this field to an HTTPRoute path matcher, we propose the following
+scheme:
+
+- If neither service nor method is supplied and the type is `Exact`, translate to a
+  path match of `/` with type `PathPrefix`
+- If service and method are supplied and the type is `Exact`, translate to a
+  path match of `/SERVICE/METHOD` with type `Exact`
+- If service is supplied but not method and the type is `Exact`, translate to a
+  path match of `/SERVICE/` with type `PathPrefix`
+- Otherwise, translate the service and the method individually into regular
+  expressions (with an empty field interpreted as `.*`), and translate to an
+  HTTPRoute path matcher of type `RegularExpression` with value `/SERVICE_REGEX/MATCHER_REGEX`
+
+
+#### An Example Implementation
+
+In the interest of proving the feasibility of implementation of this proposal,
+we include a sketch of an implementation for an xDS-based control plane.
+
+Suppose the following resources are applied to the cluster:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  gatewayClassName: acme-lb
+  listeners:
+  - name: http
+    protocol: HTTP
+    port: 80
+---
+kind: GRPCRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-grpcroute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - method:
+        service: foo.bar
+    backendRefs:
+    - name: grpc-backend
+      port: 50051
+---
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-httproute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - path:
+        type: PathPrefix
+        value: /foo
+    backendRefs:
+    - name: http-backend
+      port: 80
+```
+
+According to the above scheme, we translate the gRPC method matcher to an HTTP
+path matcher of type `PathPrefix` and value `/foo.bar/`. Following the HTTPRoute
+precedence rules, `/foo.bar/` has more characters than `/foo`, so it has higher
+precedence. Hence, the GRPCRoute's rule must be evaluated before the HTTPRoute's
+rule.
+
+Accordingly, an xDS server should create something like the following
```

Comment:
Best to remove specifics of implementation, I don't _think_ we do this anywhere else.

Reply:
I agree we haven't done this very much, but I think the example benefits from showing the eventual LB config in this case - the fact that's an xDS config isn't that important.

Reply:
This section has been removed due to a change in merge semantics.